[COMPLETED] Added express_oauth_microsoft template

GMHarish285 opened the pull request:
#32 
Added command express_oauth_microsoft.
Created template for express_oauth_microsoft.
Microsoft OAuth template will be created when user runs `qse init -t express_oauth_microsoft`.

Key points in the template:
- created a basic template using postman-microsoft and token integrity check done using jsonwebtoken
- tested using Microsoft Azure - Microsoft Entra ID

Ashrockzzz2003 commented on package.json:
Please remove the imports relevant for the template from here. It's already there in `package.json` of template and that's enough.

This is `package.json` for the tool.

GMHarish285 replied:
Got it. I also made a mistake in specifying the path for .env in index.js since I was working in the original repo and the path differs from the actual path when template is created from CMD. I will correct that too.

Ashrockzzz2003 reviewed (changes requested):
Thanks for adding the unit tests yourselves! Just one change after first pass!

Overall looks good.

GMHarish285 commented:
I removed the dependencies relevant for template from the package.json of the tool, and fixed a couple of path issues. Should i make a separate pull request for this?

Ashrockzzz2003 commented:
> I removed the dependencies relevant for template from the package.json of the tool, and fixed a couple of path issues. Should i make a separate pull request for this?

Nope
Same PR.
When you push code to the branch from which you created this PR, it'll get updated here.

GMHarish285 commented:
Can you try testing the application now?

GMHarish285 commented:
I made changes to the flow

Ashrockzzz2003 commented on templates/express_oauth_microsoft/router/app.js:
```diff
@@ -0,0 +1,14 @@
+
+// home page where user info is displayed
+router.get('/', authenticateUser, (req, res) => {
+    const user = req.user
+    const token = req.cookies?.token
+    res.send(`Your name is: ${user.name}<br>Your email is: ${user.email}<br>Your token is: ${token}`)
```
Enclose in a try catch block. Catch will throw 500 and write to a log file. Refer the `express_mysql` template for doing it.

Ashrockzzz2003 commented on templates/express_oauth_microsoft/router/auth.js:
```diff
@@ -0,0 +1,39 @@
+
+// token is verified if exists otherwise redirects to microsoft auth
+router.get('/login', (req, res) => {
+    const token = req.cookies?.token
+    if (token) {
+        try {
+            const user = jwt.verify(token, process.env.JWT_SECRET)
+            return res.redirect('/')
+        } catch {
+            res.clearCookie('token')
+        }
+    }
+    res.redirect('/auth/microsoft/redirect')
```
Enclose in a try catch block. Catch will throw 500 and write to a log file. Refer the `express_mysql` template for doing it.

Ashrockzzz2003 commented on templates/express_oauth_microsoft/router/auth.js:
```diff
@@ -0,0 +1,39 @@
+
+// microsoft auth occurs here by calling microsoft strategy middleware
+router.get('/microsoft/redirect', passport.authenticate('microsoft', { session: false }), (req, res) => {
+    if (req.user) {
+        const token = jwt.sign(req.user, process.env.JWT_SECRET, { expiresIn: '1h' })
+        res.cookie('token', token, { httpOnly: true }) // use in development
+        // res.cookie('token', token, { httpOnly: true, secure: true }) // use in production
+        res.redirect('/')
+    } else {
+        res.status(400).send('Authentication failed.')
+    }
```
Enclose in a try catch block. Catch will throw 500 and write to a log file. Refer the `express_mysql` template for doing it.

Ashrockzzz2003 commented on templates/express_oauth_microsoft/router/auth.js:
```diff
@@ -0,0 +1,39 @@
+
+// handles logout
+router.get('/logout', (req, res) => {
+    res.clearCookie('token')
+    res.redirect('/')
```
Enclose in a try catch block. Catch will throw 500 and write to a log file. Refer the `express_mysql` template for doing it.

Ashrockzzz2003 commented on templates/express_oauth_microsoft/middleware/authenticate.js:
```diff
@@ -0,0 +1,18 @@
+
+// cookie token authetication
+const authenticateUser = (req, res, next) => {
+    const token = req.cookies?.token
+    if (!token) {
+        return res.redirect('/auth/login')
+    }
+    try {
+        const user = jwt.verify(token, process.env.JWT_SECRET)
+        req.user = user
+        next()
+    } catch (error) {
+        res.status(401).send('Authentication failed. Please log in again.')
+    }
```
Write error to a log file. Refer the `express_mysql` template for doing it.

Ashrockzzz2003 reviewed (changes requested):
> [!Important]
> Please run `git pull origin main`. The `bin/configs.js` just got easier to setup. Just the name of the template is enough with the `package.json` configured inside the template.

Everything works now and flow looks good. Some improvements can be done. 

- [x] Log errors to log files. 
- [ ] Enclose all functions with a try catch. 
- [ ] Move the logic of routes to a controller file.

Refer `express_mysql` to get a better idea on how to do this.

GMHarish285 commented:
i have made the changes you asked me to. i put the error handler as a separate file as a wrapper function, so the code looks cleaner, hope that is fine.

Ashrockzzz2003 commented on templates/express_oauth_microsoft/index.js:
```diff
@@ -0,0 +1,60 @@
+            const email = profile.emails?.[0]?.value || 'No email available'
+            const name = profile.displayName || 'No name available'
+            const token = jwt.sign({ email, name }, process.env.JWT_SECRET, { expiresIn: '1h' }) // generating token
+            const redirect_url = `http://localhost:3000/user?token=${token}` // fallback page
```
unused variable?

GMHarish285 replied:
i will remove redirect_url, but i think it is better to keep token, since developer might need it at a later stage for session management or anything else

Ashrockzzz2003 commented on templates/express_oauth_microsoft/controller/authController.js:
```diff
@@ -0,0 +1,33 @@
+    const token = req.cookies?.token
+    if (token) {
+        try {
+            const user = jwt.verify(token, process.env.JWT_SECRET)
```
unused?

Ashrockzzz2003 commented on templates/express_oauth_microsoft/index.js:
```diff
@@ -0,0 +1,60 @@
+import { configDotenv } from 'dotenv'
+configDotenv({ path: "./templates/express_oauth_microsoft/.env" }) // Use when testing the tool
+// configDotenv({ path: "./.env" }) // use when testing the template
+
```
not necessary. comment this one out. uncomment the below line. ideally when the template is generated, it should work.

Ashrockzzz2003 reviewed (approved):
LGTM

Ashrockzzz2003 commented:
/bounty 70 @GMHarish285

Ashrockzzz2003 commented:
Awesome work
Merging it @GMHarish285
